### PR TITLE
add /emote support to talkinghead

### PR DESCRIPTION
**Pull request by Technologicat**
Add `/emote xxx` support for talkinghead (at least for now, only when VN mode is off; that's something I'd like to discuss the pros/cons of).

This depends on https://github.com/SillyTavern/SillyTavern-Extras/pull/213 for backend support, as it needs the new API endpoint `/api/talkinghead/set_emotion`.

**NOTE**: This is a quick prototype. Due to some validation code in `setSpriteSlashCommand`, there currently **must** be a static sprite for the given expression name for that expression name to be detected as valid. So beside `talkinghead.png`, your character also needs the static sprite for emotion `"xxx"` for `"/emote xxx"` to work, although `talkinghead` doesn't use the static sprites.

My suggestion is that, before making this PR final, we could check `extension_settings.expressions.talkinghead` in `setSpriteSlashCommand`, and skip the validation check in `talkinghead` mode. There is a check at the server end to default to `"neutral"` (and log a warning) if the expression name sent by the client is not recognized.

As for the code, I'm modding this [as a pythonista](https://github.com/Technologicat/js-for-pythonistas). `async/await` syntax would read much nicer, but we're inside an arrow function that handles a response from a promise, so I suppose that in order to use `async/await` here, we would also need to convert the call to `talkingHeadCheck` to use `async/await` instead of bare promises.

Comments?

### CODE REVIEW

File: public/scripts/extensions/expressions/index.js [hunk 1]
```diff
@@ -1113,14 +1113,29 @@ async function setExpression(character, expression, force) {
...
-                //console.log("searching");
-                if (imgElement && imgElement instanceof HTMLImageElement) {
-                    //console.log("setting value");
-                    imgElement.src = getApiUrl() + '/api/talkinghead/result_feed';
-                }
-
+                // Set the talkinghead emotion to the specified expression
+                // TODO: For now, talkinghead emote only supported when VN mode is off; see also updateVisualNovelMode.
+                const url = new URL(getApiUrl());
+                url.pathname = '/api/talkinghead/set_emotion';
```
Comment (Cohee1207) [inline public/scripts/extensions/expressions/index.js:1118]: This doesn't have backward compatibility with older revisions of extras. I'd like to make a call of set_emotion not required here

Comment (Cohee1207) [reply public/scripts/extensions/expressions/index.js:1118]: TL;DR: replace then on line 1127 with finally, should be enough

**Comment by Cohee1207** [comment]
You also can use async/await inside of arrow functions if you mark the [arrow function as async](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions#syntax)

File: public/scripts/extensions/expressions/index.js (hunk 1 above)
Comment (Technologicat) [reply public/scripts/extensions/expressions/index.js:1118]: True, good point. So allowing ‘set_emotion‘ to 404 against previous versions is the preferred way to handle that?

Ok, I'll revise the code.

Comment (Cohee1207) [reply public/scripts/extensions/expressions/index.js:1118]: Yes. Even if 404 - it is not mission-critical and could be skipped.

**Comment by Technologicat** [comment]
> You also can use async/await inside of arrow functions if you mark the [arrow function as async](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions#syntax)

Nice, thanks!

I'll do that.

**Comment by Technologicat** [comment]
Here, fixed. I think this patch should now fulfill both of our code quality standards :)

**Some questions** before declaring this final:

- **The static sprite check in `setSpriteSlashCommand` should be disabled in `talkinghead` mode**, because `talkinghead` doesn't use the static sprites (other than `talkinghead.png`), and the backend will validate the emotion name anyway. **I could change this - are you ok with the change?**
  - As of https://github.com/SillyTavern/SillyTavern-Extras/pull/213, the backend always replies 200 OK, even if the specified emotion does not exist (in that case the backend defaults to `neutral`). I could add error reporting for that in the next PR for the backend.

- **Should I add a `get_emotion` endpoint for future use?** OTOH, YAGNI, but IMHO having only a setter (like we do now) is almost never the Right Thing. It's foreseeable that the client may need to query the current emotion reliably.

- **Visual novel mode in *Character expressions*.** Do we want to support `/emote` for `talkinghead` in VN mode? I'd like your input on this, because I don't use the VN mode myself. The updater function `updateVisualNovelMode` seems to contain a lot of logic related to group chats, which I also don't use (at least so far).
  - The `talkinghead` view of the character is needs a solid edge to "sit" on. The obvious solutions are at the bottom of the window, or above the top edge of a non-translucent chat panel.
  - Currently the backend is designed to run only one global instance of the animator. If we want more, we need to change the API as well as the backend implementation.
  - Current GPUs don't have enough compute to run many talkingheads at interactive framerates simultaneously, anyway. But this may change as more powerful GPUs become available.

- **I have planned a couple of other small fixes that should be simple to implement.** Wait for them here, or create a new PR once done?

**Comment by Cohee1207** [comment]
> The static sprite check in setSpriteSlashCommand should be disabled in talkinghead mode

Ok, /emote should call set_emotion instead of picking sprite in talkinghead mode. I agree. You can make it a separate PR.

>  Do we want to support /emote for talkinghead in VN mode

VN mode for groups is for groups only, I don't think it has any use for talkinghead right now. So no.

> 2. Should I add a get_emotion endpoint for future use?

Likely not. Only the frontend state matters.

**Pull request merged**